sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+      * `0x02` means this is a pay-to-witness-public-key-hash-over-pay-to-script-hash (p2wpkh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wpkh witness script
+      * `0x03` means this is a pay-to-witness-script-hash-over-pay-to-script-hash (p2wsh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wsh witness script
+      * `0x04` means this is a pay-to-witness-public-key-hash (p2wpkh) address, and `hashbytes` is the 20-byte hash160 of the witness script
+      * `0x05` means this is a pay-to-witness-script-hash (p2wsh) address, and `hashbytes` is the 32-byte sha256 of the witness script
+      * `0x06` means this is a pay-to-taproot (p2tr) address, and `hashbytes` is the 32-byte sha256 of the witness script
+
+**Rationale:**  This method empowers developers to query Bitcoin state.  Stacks
+2.1 adds the ability for smart contracts to query PoX payouts directly, with applications
+for Stacking pools and other Stacking-centric programs.
+
+**Examples:**
+
+```clarity
+(get-burn-block-info? header-hash u677050) ;; Returns (some 0xe67141016c88a7f1203eca0b4312f2ed141531f59303a1c267d7d83ab6b977d8)
+(get-burn-block-info? pox-addrs u677050) ;; Returns (some (tuple (addrs ((tuple (hashbytes 0x395f3643cea07ec4eec73b4d9a973dcce56b9bf1) (version 0x00)) (tuple (hashbytes 0x7c6775e20e3e938d2d7e9d79ac310108ba501ddb) (version 0x01)))) (payout u123)))
+```
+
+### New method: `slice`
+
+* **Input Signature:** `(slice (sequence sequence_A) (left-position uint) (right-position uint))`
+* **Output Signature:** `(optional sequence_A)`
+
+The `slice` function attempts to return a sub-sequence of `sequence_A` that starts
+at `left-position` (inclusive), and ends at `right-position`
+(non-inclusive).
+
+If `left-position`==`right-position`, the function returns an empty
+sequence.
+
+If either `left-position` or `right-position` are out of bounds OR if
+`right-position` is less than `left-position`, the function returns
+`none`.
+
+Values in `sequence_A` are zero-based.  That is, the first item in `sequence_A`
+is at index 0. 
+
+**Rationale:** This method facilitates parsing user-supplied encoded data, such
+as data from oracles and other off-chain services.  While a variant of `slice`
+could be implemented entirely in Clarity, it would be much slower and much more
+expensive than supplying a native method.
+
+**Examples:**
+
+```clarity
+(slice "blockstack" u5 u10) ;; Returns (some "stack")
+(slice (list 1 2 3 4 5) u5 u9) ;; Returns none
+(slice (list 1 2 3 4 5) u3 u4) ;; Returns (some (4))
+(slice "abcd" u0 u2) ;; Returns (some "ab")
+(slice "abcd" u1 u3) ;; Returns (some "bc")
+(slice "abcd" u2 u2) ;; Returns (some "")
+(slice "abcd" u3 u1) ;; Returns none
+```
+
+### New method: `string-to-int`
+
+* **Input Signature:** `(string-to-int (input (string-ascii|string-utf8)))`
+* **Output Signature:** `(optional int)`
```

Comment:
I'm going to comment only this function but it applies to all 4 (`string-to-int`, `string-to-uint`, `int-to-asci`, `int-to-utf8`).
If target data type is specific, then source data type should be specific as well and instead of 4 we should have 8 functions:
 - `ascii-to-int`
 - `utf8-to-int`
 - `asci-to-uint`
 - `utf8-to-uint`
 - `ascii-to-int`
 - `utf8-to-int`
 - `ascii-to-uint`
 - `utf8-to-uint`

Especially when we `int-to-utf8` and `int-to-ascii` suggests that input data type is `int` while in fact it accepts both `uint` and `int`

Reply:
> If target data type is specific, then source data type should be specific as well 

Why?

Reply:
To make it explicit what is the data type from which you are converting. It will make code analysis much simpler.
If I squint hard enough I would probably accept `string-to-X` but `int-to-X` suggests that input type should be `int` while in fact it accepts both `uint` and `int, so `number-to-X` would fit better.

Also it would be great to add max length of input of `string-to-X` and output of `number-to-X` as they are known and should be enforced and respected at the contract validation level.

Reply:
This exact topic came up at one of the blockchain meetings back when this was being implemented (I don't remember who brought it up). The group considered `number`, but thought that developers might think that included floating/fixed point numbers; we considered `integer`, but thought that just made it longer to type and didn't really clarify anything. Personally, I would be okay with `number`.